File: data/scripts/motz.py
#http://rosalind.info/problems/motz/

import sys

MODULO = 1000000
COMPLEMENTI = {"A": "U", "U": "A", "C": "G", "G": "C"}

def lettura(filename):
    with open(filename) as f:
        record = ""
        first = True
        for riga in f:
            riga = riga.rstrip("\n")
            if riga[0] == ">":
                if not first:
                    return record
                first = False
            else:
                record += riga
    return record

def conta_matching(s, memo):
    n = len(s)
    if n in (0, 1):
        return 1
    if s in memo:
        return memo[s]
    totale = conta_matching(s[1:], memo)
    for k in range(1, n):
        if COMPLEMENTI.get(s[0]) == s[k]:
            totale += conta_matching(s[1:k], memo) * conta_matching(s[k + 1:], memo)
            totale %= MODULO
    memo[s] = totale
    return totale

def main():
    sys.setrecursionlimit(10000)
    s = lettura("dataset.txt")
    memo = {}
    print(conta_matching(s, memo))

if __name__ == "__main__":
    # execute only if run as a script
    main()

File: public/js-solutions/motz.mjs
// Motzkin Numbers and RNA Secondary Structures (Rosalind ID: MOTZ) -
// soluzione JavaScript indipendente, non una trascrizione di
// problem.py: stessa logica (ricorsione con memoizzazione sul numero di
// possibili strutture secondarie non-crossing, mod 1.000.000),
// riscritta in modo idiomatico per JS.
//
// Contratto: riceve il contenuto testuale di dataset.txt, restituisce
// l'output testuale (stessa forma dell'output Python).
const MODULO = 1000000;
const COMPLEMENTI = { A: "U", U: "A", C: "G", G: "C" };

function primaSequenzaFasta(testo) {
  const righe = testo.split("\n").map((r) => r.replace(/\r$/, ""));
  let record = "";
  let first = true;
  for (const riga of righe) {
    if (riga.startsWith(">")) {
      if (!first) return record;
      first = false;
    } else {
      record += riga;
    }
  }
  return record;
}

function contaMatching(s, memo) {
  const n = s.length;
  if (n === 0 || n === 1) return 1;
  if (memo.has(s)) return memo.get(s);

  let totale = contaMatching(s.slice(1), memo);
  for (let k = 1; k < n; k++) {
    if (COMPLEMENTI[s[0]] === s[k]) {
      totale += contaMatching(s.slice(1, k), memo) * contaMatching(s.slice(k + 1), memo);
      totale %= MODULO;
    }
  }
  memo.set(s, totale);
  return totale;
}

export default function solve(datasetText) {
  const s = primaSequenzaFasta(datasetText);

  if (!s) {
    throw new Error("Input non valido: nessuna sequenza FASTA trovata");
  }

  const memo = new Map();
  return `${contaMatching(s, memo)}\n`;
}
